### PR TITLE
refactor and fix: addressing recent changes

### PR DESCRIPTION
**Pull request by kupermind**
- Addressing recent changes.

### CODE REVIEW

File: contracts/BalanceTrackerFixedPriceNative.sol [hunk 1]
```diff
@@ -35,7 +35,14 @@ contract BalanceTrackerFixedPriceNative is BalanceTrackerFixedPriceBase {
         wrappedNativeToken = _wrappedNativeToken;
     }
 
-    function _checkNativeValue() internal virtual override {}
+    function _getOrRestrictNativeValue() internal virtual override returns (uint256) {
```
Comment (kupermind) [inline contracts/BalanceTrackerFixedPriceNative.sol:38]: This is the current refactor of incoming payment.

File: contracts/BalanceTrackerFixedPriceToken.sol [hunk 2]
```diff
@@ -38,16 +43,29 @@ contract BalanceTrackerFixedPriceNative is BalanceTrackerFixedPriceBase {
...
         if (msg.value > 0) {
             revert NoDepositAllowed(msg.value);
         }
+
+        return 0;
     }
 
     function _getRequiredFunds(address requester, uint256 balanceDiff) internal virtual override returns (uint256) {
-        // TODO check balances before and after
+        uint256 balanceBefore = IToken(olas).balanceOf(address(this));
```
Comment (kupermind) [inline contracts/BalanceTrackerFixedPriceToken.sol:56]: Additional check, although with OLAS it's pretty much safe by default.

File: contracts/MechFactoryFixedPriceNative.sol [hunk 3]
```diff
@@ -37,12 +50,19 @@ contract MechFactoryFixedPriceNative {
         // Decode max delivery rate
         uint256 maxDeliveryRate = abi.decode(payload, (uint256));
 
+        uint256 localNonce = _nonce;
         // Get salt
-        bytes32 salt = keccak256(abi.encode(block.timestamp, msg.sender, serviceId));
+        bytes32 salt = keccak256(abi.encode(block.timestamp, msg.sender, serviceId, localNonce));
```
Comment (kupermind) [inline contracts/MechFactoryFixedPriceNative.sol:55]: Adding nonce as per audit suggestion

File: contracts/MechFixedPriceToken.sol [hunk 4]
```diff
@@ -5,13 +5,16 @@ import {OlasMech} from "./OlasMech.sol";
 
 /// @title MechFixedPriceToken - Smart contract for OlasMech that accepts a fixed price payment for services in native token.
 contract MechFixedPriceToken is OlasMech {
+    // keccak256(FixedPriceToken) = 3679d66ef546e66ce9057c4a052f317b135bc8e8c509638f7966edfd4fcf45e9
+    bytes32 public constant PAYMENT_TYPE = 0x3679d66ef546e66ce9057c4a052f317b135bc8e8c509638f7966edfd4fcf45e9;
```
Comment (kupermind) [inline contracts/MechFixedPriceToken.sol:9]: Hashing approach always wins :). Although in this case it feels like we maybe need to call it FixedPriceOlas? Either way is fine.

File: contracts/MechMarketplace.sol [hunk 5]
```diff
@@ -468,24 +467,24 @@ contract MechMarketplace is IErrorsMarketplace {
         // Decrease the number of undelivered requests
         numUndeliveredRequests--;
         // Increase the amount of requester delivered requests
-        mapDeliveryCounts[requester]++;
+        mapDeliveryCounts[mechDelivery.requester]++;
         // Increase the amount of mech delivery counts
-        mapAgentMechDeliveryCounts[msg.sender]++;
+        mapMechDeliveryCounts[msg.sender]++;
         // Increase the amount of mech service multisig delivered requests
         mapMechServiceDeliveryCounts[mechServiceMultisig]++;
```
Comment (kupermind) [inline contracts/MechMarketplace.sol:474]: We still need both - one is for mech statistics, another is for multisig activity tracker.

File: contracts/MechMarketplace.sol [hunk 6]
```diff
@@ -602,41 +593,5 @@ contract MechMarketplace is IErrorsMarketplace {
             }
         }
     }
-
-    /// @dev Gets the requests count for a specific account.
-    /// @param account Account address.
-    /// @return Requests count.
-    function getRequestsCount(address account) external view returns (uint256) {
```
Comment (kupermind) [inline contracts/MechMarketplace.sol:609]: Default ABI is enough

File: contracts/MechFixedPriceToken.sol (hunk 4 above)
Comment (DavidMinarsch) [reply contracts/MechFixedPriceToken.sol:9]: hm I think the token can be different kinds for the same factory contract (different factory instance) no?

**Review by DavidMinarsch — APPROVED** [review]

**Review by Sfgangloff — APPROVED** [review]

File: contracts/MechFixedPriceToken.sol (hunk 4 above)
Comment (kupermind) [reply contracts/MechFixedPriceToken.sol:9]: Makes sense, maybe encode it on a level of factory? There we can combine `FixedPriceToken` + token address or anything else.

Comment (DavidMinarsch) [reply contracts/MechFixedPriceToken.sol:9]: Hm, I'll think about it.

**Pull request merged**